typescript/packages/common-html/src/render.ts

```diff
@@ -0,0 +1,417 @@
+import { effect } from "./reactive.js";
+
+export { setDebug } from "./log.js";
+
+/** Render template with replacements */
+export const render = (tpl: Template) => {
+  if (!isTemplate(tpl)) {
+    throw new TypeError(`Expected a template object`);
+  }
+  const { template, context } = tpl;
+
+  // Create a cancel bag for gathering cancels related to reactive
+  // bindings on this element.
+  const [cancelAll, addCancel] = useCancelGroup();
+
+  // Render cached string to template
+  const fragment = renderCachedTemplate(
+    flattenTemplate(template)
+  ) as CancellableElement;
+  fragment.cancel = cancelAll;
+
+  for (const node of walkElementAndTextNodes(fragment)) {
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const element = node as Element;
+
+      // Prune disallowed elements
+      if (!isElementAllowed(element)) {
+        debug(`Element not allowed`, element);
+        element.remove();
+        continue;
+      }
+
+      // Capture attrs, since we're about to modify them during iteration.
+      const attrs = Array.from(element.attributes);
+      for (const attr of attrs) {
+        if (isPropKey(attr.name)) {
+          const cancel = bindProp(element, attr.name, attr.value, context);
+          addCancel(cancel);
+        } else if (isEventKey(attr.name)) {
+          const cancel = bindEvent(element, attr.name, attr.value, context);
+          addCancel(cancel);
+        } else {
+          const cancel = bindAttr(element, attr.name, attr.value, context);
+          addCancel(cancel);
+        }
+      }
+    } else if (node.nodeType === Node.TEXT_NODE) {
+      const cancel = bindContent(node, context);
+      addCancel(cancel);
+    }
+  }
+  debug("Rendered template", fragment);
+  return fragment;
+};
+
+export type CancellableElement = Element & Cancellable;
+
+const alwaysAllow = () => true;
+
+export type ElementSanitizer = (node: Element) => boolean;
+
+export type PropSanitizer = (
+  node: Node,
+  key: string,
+  value: unknown
+) => boolean;
+
+export type AttrSanitizer = (node: Node, key: string) => boolean;
+
+export type EventSanitizer = (node: Node, key: string) => boolean;
+
+/** The active sanitizers */
+let isElementAllowed: ElementSanitizer = alwaysAllow;
+let isPropAllowed: PropSanitizer = alwaysAllow;
+let isAttrAllowed: AttrSanitizer = alwaysAllow;
+let isEventAllowed: EventSanitizer = alwaysAllow;
+
+/**
+ * Set the property sanitizer
+ * @example
+ * import { setPropSanitizer } from "curly";
+ *
+ * setSanitizer({ ... })
+ */
+export const setSanitizer = ({
+  element,
+  prop,
+  attr,
+  event,
+}: {
+  element: ElementSanitizer;
+  prop: PropSanitizer;
+  attr: AttrSanitizer;
+  event: EventSanitizer;
+}) => {
+  isElementAllowed = element;
+  isPropAllowed = prop;
+  isAttrAllowed = attr;
+  isEventAllowed = event;
+};
+
+/** Generates a random number with 8 digits */
+const randomSalt = () => Math.random().toFixed(9).slice(2);
```

Comment:
Is this for security reasons or just to reduce the chance of accidental collisions?

Since the context is created with and associated with the template, is there really something disallowed code pull off by guessing the numbers?

In any case, the templates's hash can't be inserted into the template, so that would be a safe choice.

(Mostly trying to understand)

Reply:
My rationale here basically boils down to trying to reduce the possibility of failure.

Before we were using code to construct markup, and could enforce invariants through that code so that it was impossible to construct broken markup. Now that we're parsing a string (or rather allowing the browser to parse it), we have to be a bit clever. We walk over the resulting markup, and so can guarantee it is correct on the way out. However, the template holes are one place where it might be possible to break the render by faking a hole placeholder.

The random salt prevents collision (this is also what Lit does under the hood). I'm not sure if it is a hard security requirement to do better than this. If it is, I like the idea of using the hash, as you mentioned.